api/types/audit.go

```diff
@@ -0,0 +1,258 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"time"
+
+	"github.com/gravitational/teleport/api/defaults"
+
+	"github.com/gravitational/trace"
+)
+
+// ClusterAuditConfig defines cluster-wide audit log configuration. This is
+// a configuration resource, never create more than one instance of it.
+type ClusterAuditConfig interface {
+	Resource
+
+	// Type gets the audit backend type.
+	Type() string
+	// SetType sets the audit backend type.
+	SetType(string)
+
+	// Region gets a cloud provider region.
+	Region() string
+	// SetRegion sets a cloud provider region.
+	SetRegion(string)
+
+	// ShouldUploadSessions returns whether audit config
+	// instructs server to upload sessions.
+	ShouldUploadSessions() bool
+
+	// AuditSessionsURI gets the audit sessions URI.
+	AuditSessionsURI() string
+	// SetAuditSessionsURI sets the audit sessions URI.
+	SetAuditSessionsURI(string)
+
+	// AuditEventsURIs gets the audit events URIs.
+	AuditEventsURIs() []string
+	// SetAuditEventsURIs sets the audit events URIs.
+	SetAuditEventsURIs([]string)
+
+	// EnableContinuousBackups is used to enable (or disable) PITR (Point-In-Time Recovery).
+	EnableContinuousBackups() bool
+	// EnableAutoScaling is used to enable (or disable) auto scaling policy.
+	EnableAutoScaling() bool
+	// ReadMaxCapacity is the maximum provisioned read capacity.
+	ReadMaxCapacity() int64
+	// ReadMinCapacity is the minimum provisioned read capacity.
+	ReadMinCapacity() int64
+	// ReadTargetValue is the ratio of consumed read to provisioned capacity.
+	ReadTargetValue() float64
+	// WriteMaxCapacity is the maximum provisioned write capacity.
+	WriteMaxCapacity() int64
+	// WriteMinCapacity is the minimum provisioned write capacity.
+	WriteMinCapacity() int64
+	// WriteTargetValue is the ratio of consumed write to provisioned capacity.
+	WriteTargetValue() float64
+}
+
+// NewClusterAuditConfig is a convenience method to to create ClusterAuditConfigV2.
+func NewClusterAuditConfig(spec ClusterAuditConfigSpecV2) (ClusterAuditConfig, error) {
+	auditConfig := ClusterAuditConfigV2{
+		Kind:    KindClusterAuditConfig,
+		Version: V2,
+		Metadata: Metadata{
+			Name:      MetaNameClusterAuditConfig,
+			Namespace: defaults.Namespace,
+		},
+		Spec: spec,
+	}
+
+	if err := auditConfig.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &auditConfig, nil
+}
+
+// DefaultClusterAuditConfig returns the default audit log configuration.
+func DefaultClusterAuditConfig() ClusterAuditConfig {
+	config, _ := NewClusterAuditConfig(ClusterAuditConfigSpecV2{})
```

Comment:
I think we should either make these Default functions panic on error from `CheckAndSetDefaults`, or (preferably) add basic test coverage to make sure the default passes `CheckAndSetDefaults`. Some resources won't pass validation with an empty spec, and these resources might be updated in the future to be the same. Doesn't need to be in this PR, just curious what you think and we can add it later.

Reply:
I agree this is not an ideal solution. Even though personally I'd be inclined to panic here as the most straightforward solution, that approach was rejected in https://github.com/gravitational/teleport/pull/5159#discussion_r545477156.

For reference, the current approach was first adopted because of https://github.com/gravitational/teleport/pull/6638#discussion_r625294476.

In conclusion, adding tests to ensure all the `Default*` methods return valid resources as per your suggestion seems the only way to go.